crates/sui-framework/packages/sui-framework/sources/token.move

```diff
@@ -0,0 +1,731 @@
+    /// `TokenPolicyCap` to decrease supply. For scenarios like this a
+    /// `TreasuryCap` is required (see `confirm_with_treasury_cap`).
+    public fun confirm_with_policy_cap<T>(
+        _policy_cap: &TokenPolicyCap<T>,
+        request: ActionRequest<T>,
+        _ctx: &mut TxContext
+    ): (String, u64, address, Option<address>) {
+        assert!(option::is_none(&request.spent_balance), ECantConsumeBalance);
+
+        let ActionRequest {
+            name, amount, sender, recipient, approvals: _, spent_balance
+        } = request;
+
+        option::destroy_none(spent_balance);
+
+        (name, amount, sender, recipient)
+    }
+
+    /// Confirm an `ActionRequest` as the `TreasuryCap` owner. This function
+    /// allows `TreasuryCap` owner to perform Capability-gated actions ignoring
+    /// the ruleset specified in the `TokenPolicy`.
+    ///
+    /// Unlike `confirm_with_policy_cap` this function allows `spent_balance`
+    /// to be consumed, decreasing the `total_supply` of the `Token`.
+    public fun confirm_with_treasury_cap<T>(
+        treasury_cap: &mut TreasuryCap<T>,
+        request: ActionRequest<T>,
+        _ctx: &mut TxContext
+    ): (String, u64, address, Option<address>) {
+        let ActionRequest {
+            name, amount, sender, recipient, approvals: _,
+            spent_balance
+        } = request;
+
+        if (option::is_some(&spent_balance)) {
+            balance::decrease_supply(
+                coin::supply_mut(treasury_cap),
+                option::destroy_some(spent_balance)
+            );
+        } else {
+            option::destroy_none(spent_balance);
+        };
+
+        (name, amount, sender, recipient)
+    }
+
+    // === Rules API ===
+
+    /// Add an "approval" to the `ActionRequest` by providing a Witness.
+    /// Intended to be used by Rules to add their own approvals, however, can
+    /// be used to add arbitrary approvals to the request (not only the ones
+    /// required by the `TokenPolicy`).
+    public fun add_approval<T, W: drop>(
+        _t: W, request: &mut ActionRequest<T>, _ctx: &mut TxContext
+    ) {
+        vec_set::insert(&mut request.approvals, type_name::get<W>())
+    }
+
+    /// Add a `Config` for a `Rule` in the `TokenPolicy`. Rule configuration is
+    /// independent from the `TokenPolicy.rules` and needs to be managed by the
+    /// Rule itself. Configuration is stored per `Rule` and not per `Rule` per
+    /// `Action` to allow reuse in different actions.
+    ///
+    /// - Rule witness guarantees that the `Config` is approved by the Rule.
+    /// - `TokenPolicyCap` guarantees that the `Config` setup is initiated by
+    /// the `TokenPolicy` owner.
+    public fun add_rule_config<T, Rule: drop, Config: store>(
+        _rule: Rule,
+        self: &mut TokenPolicy<T>,
+        cap: &TokenPolicyCap<T>,
+        config: Config,
+        _ctx: &mut TxContext
+    ) {
+        assert!(object::id(self) == cap.for, ENotAuthorized);
```

Comment:
Since we're using this assertion multiple times for access control, should we move it to a function with a more descriptive name?

Reply:
Then you'd either have a function `assert_cap_matches(policy, cap)` which is not great as this function is not very reusable. Or redo an assertion everywhere as `assert!(cap_matches(policy, cap), ENotAuthorized)`. Both feel like an overkill for something that simple. At least for now.